src/storage/src/upsert_continual_feedback.rs

```diff
@@ -0,0 +1,826 @@
+
+        let mut error_emitter = (&mut health_output, &health_cap);
+        let mut legacy_errors_to_correct = vec![];
+
+
+        loop {
+            tokio::select! {
+                Some(persist_event) = persist_input.next() => {
+
+                    // Buffer as many events as possible. This should be
+                    // bounded, as new data can't be produced in this worker
+                    // until we yield to timely.
+                    let persist_events = [persist_event]
+                        .into_iter()
+                        .chain(std::iter::from_fn(|| persist_input.next().now_or_never().flatten()));
+
+                    // Read away as much input as we can.
+                    for persist_event in persist_events {
+                        tracing::trace!(?persist_event, "persist input");
+
+                        match persist_event {
+                            AsyncEvent::Data(_cap, data) => {
+                                persist_stash.extend(data.into_iter().map(|((key, value), ts, diff)| {
+                                    (key, value, ts, diff)
+                                }))
+                            }
+                            AsyncEvent::Progress(upper) => persist_upper = upper,
+                        }
+                    }
+
+                    if hydrating {
+
+                        // Determine if this is the last time we call
+                        // consolidate_snapshot_chunk, and update our
+                        // `hydrating` state.
+                        //
+                        // There are two situations in which we're reading a
+                        // snapshot:
+                        //
+                        // 1. Reading the initial snapshot from the source: in
+                        //    this case there is no state in our output yet that
+                        //    we could read back in and our resume_upper is [0].
+                        //    We know we're done once the persist_upper is
+                        //    "past" that.
+                        // 2. Re-hydrating our state from persist: some instance
+                        //    of the source has already read/ingested/written
+                        //    down the initial snapshot from the source. We are
+                        //    reading a snapshot of our own upsert state back in
+                        //    from persist. In this case we read the snapshot at
+                        //    a timestamp that is "before" the resume upper, and
+                        //    we know that we're done when the persist_upper is
+                        //    "at or past" the resume_upper.
+                        //
+                        // It's unfortunate that we have to differentiate
+                        // between these cases, both here and in code below. The
+                        // root of the problem is that when reading the initial
```

Comment:
I didn't understand the problem. If `resume_upper == [0]` then I would expect that we immediately set `hydrating = false` and move to the other phase of the operator. Why do we have to differentiate between the two cases?

Reply:
You're right, this version of the operator wouldn't need that. I kept it like this because a) the classic upsert operator also went through it's snapshot/hydration logic, even when the resume upper is zero, and b) it _will_ be needed once we emit provisional updates when ingesting the first source snapshot. I have a follow-up PR for that already, which passes ci/nightly as well, but wanted to keep this PR simpler. It's related to what you're saying here: https://github.com/MaterializeInc/materialize/pull/29718#discussion_r1820626598.

Also, if this gets to complicated we can do a call/zoom